File: setu/matching/semantic.py
"""Free-text semantic matching (step 4 of the cascade).

Family narratives ("wearing a green saree, lost near the blue water tank") and
volunteer notes rarely share exact words but often share meaning. We embed the
free-text of each record and use cosine similarity as an extra, modest signal.

Production: sentence embeddings stored in pgvector with an HNSW index, queried
1-to-few. The CONTRACT is: `embed(text) -> vector`, similarity = cosine, top-k by
ANN. Here `embed()` is a runnable, dependency-free hashed character-n-gram
bag-of-features (clearly a MOCK of a real multilingual embedder) so the demo runs
anywhere; swapping in a real model means replacing `embed()` only.

IMPORTANT: physical_description is QUARANTINED out of this signal by default —
the data shows it is unreliable. Only intentional free_text context is embedded.
The semantic contribution to the score is capped low (see cascade.py).
"""

from __future__ import annotations

import hashlib
import math
import re

DIM = 256  # mock embedding dimensionality


def _tokens(text: str) -> list[str]:
    text = text.lower()
    words = re.findall(r"[a-zऀ-ॿ]+", text)
    grams = []
    for w in words:
        grams.append(w)
        for i in range(len(w) - 2):  # char trigrams for fuzzy overlap
            grams.append(w[i:i + 3])
    return grams


def embed(text: str | None) -> list[float]:
    """MOCK multilingual embedder. Hashed-n-gram bag, L2-normalised.
    Same contract as a real embedder: str -> fixed-dim unit vector."""
    vec = [0.0] * DIM
    if not text:
        return vec
    for tok in _tokens(text):
        h = int(hashlib.md5(tok.encode("utf-8")).hexdigest(), 16)
        idx = h % DIM
        sign = 1.0 if (h >> 8) & 1 else -1.0
        vec[idx] += sign
    norm = math.sqrt(sum(v * v for v in vec))
    if norm > 0:
        vec = [v / norm for v in vec]
    return vec


def cosine(a: list[float], b: list[float]) -> float:
    return sum(x * y for x, y in zip(a, b))


def semantic_similarity(text_a: str | None, text_b: str | None) -> float:
    """Cosine similarity in [-1, 1] (typically [0, 1] for these vectors).
    Returns 0.0 if either side has no free text."""
    if not text_a or not text_b:
        return 0.0
    return round(cosine(embed(text_a), embed(text_b)), 4)

File: web/lib/setu/semantic.ts
// Free-text semantic matching (port of semantic.py).
// Runnable hashed-n-gram embedding (MOCK of a real multilingual embedder);
// the contract is text -> unit vector, similarity = cosine.

const DIM = 256;

function tokens(text: string): string[] {
  const lower = text.toLowerCase();
  const words = lower.match(/[a-zऀ-෿]+/g) ?? [];
  const grams: string[] = [];
  for (const w of words) {
    grams.push(w);
    for (let i = 0; i < w.length - 2; i++) grams.push(w.slice(i, i + 3));
  }
  return grams;
}

// FNV-1a 32-bit hash (deterministic, no deps).
function hash32(s: string): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

export function embed(text: string | null | undefined): number[] {
  const vec = new Array(DIM).fill(0);
  if (!text) return vec;
  for (const tok of tokens(text)) {
    const h = hash32(tok);
    const idx = h % DIM;
    const sign = (h >> 8) & 1 ? 1 : -1;
    vec[idx] += sign;
  }
  let norm = 0;
  for (const v of vec) norm += v * v;
  norm = Math.sqrt(norm);
  if (norm > 0) for (let i = 0; i < DIM; i++) vec[i] /= norm;
  return vec;
}

export function cosine(a: number[], b: number[]): number {
  let dot = 0;
  for (let i = 0; i < a.length; i++) dot += a[i] * b[i];
  return dot;
}

export function semanticSimilarity(
  a: string | null | undefined,
  b: string | null | undefined,
): number {
  if (!a || !b) return 0;
  const sim = cosine(embed(a), embed(b));
  return Math.round(sim * 10000) / 10000;
}
